sdk/typespec/typespec_derive/src/lib.rs

```diff
@@ -0,0 +1,205 @@
+extern crate proc_macro;
+
+use proc_macro2::{Span, TokenStream};
+use syn::parse::ParseStream;
+use syn::spanned::Spanned;
+use syn::{Attribute, DeriveInput, LitStr, Meta, Path};
+
+/// Derive macro for implementing `Model` trait.
+/// Deriving this trait allows a type to be deserialized from an HTTP response body.
+/// By default, the type must also implement `serde::Deserialize`, or the generated code will not compile.
+///
+/// ## Attributes
+///
+/// The following attributes are supported on the struct itself:
+///
+/// ### `#[typespec(crate)]`
+///
+/// The 'crate' attribute specifies an alternate module path, other than the default of `typespec_client_core`, to reference the typespec client crate.
+///
+/// ```rust
+/// # use typespec_derive::Model;
+/// # use serde::Deserialize;
+/// extern crate typespec_client_core as my_typespec;
+///
+/// #[derive(Model, Deserialize)]
+/// #[typespec(crate = "my_typespec")]
+/// struct MyModel {
+///   value: String
+/// }
+/// ```
+///
+/// ### `#[typespec(format)]`
+///
+/// The format attribute specifies the format of the response body. The default is `json`.
+/// If compiling with the `xml` feature, the value `xml` is also supported.
+///
+/// ```rust
+/// # use typespec_derive::Model;
+/// # use serde::Deserialize;
+/// #[derive(Model, Deserialize)]
+/// #[typespec(format = "xml")]
+/// struct MyModel {
+///   value: String
+/// }
+/// ```
+///
+/// **NOTE:** Using formats other than JSON may require enabling additional features in `typespec_client_core`.
+#[proc_macro_derive(Model, attributes(typespec))]
+pub fn model_derive(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    let ast: DeriveInput = syn::parse(input).unwrap();
+    let body = generate_body(ast);
```

Comment:
I eventually want to add more derive macros (lower pri right now, but like for extensible enums that have a lot of boiler plate), so I'd use a function name indicative of the derive macro e.g., `generate_model` or `model_derive_impl`. I also recommend considering putting it in a separate module since, when we add more macros, this `lib.rs` will become monolithic. Makes it easier to segment specific derives. I know serde et. al. do that...granted, they have much larger impls as well.

Reply:
Done. The proc macro function is required to live in the root of the crate, but I can organize it into an inner function in a trait-specific module that actually does all the work.

Reply:
Yeah, and that's what serde does as well. Their `src/lib.rs` is basically just calling all the functions in separate modules. Nice and neat.